### PR TITLE
feat(planner): support more aggregate syntax

### PR DESCRIPTION
**Pull request by xudong963**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Support the following syntax
- `select sum(a) from t;`
- `select count(a) from t;`
- `select count() from t`;
- `select count(1) from t`;
-  `select count(*) from t`;
- `select count() from t group by a`;
- `select count(1) from t group by a`;

## Changelog

- New Feature

## Related Issues

Fixes #issue

### CODE REVIEW

File: query/src/sql/planner/binder/scalar.rs [hunk 1]
```diff
@@ -160,6 +132,80 @@ impl ScalarBinder {
...
+                scalar_binder
+                    .bind_expr(arg, bind_context)?
+                    .as_any()
+                    .downcast_ref::<Scalar>()
+                    .ok_or_else(|| ErrorCode::UnImplement("Can't downcast to Scalar"))?
+                    .clone(),
+            );
+        }
+
+        let col_pairs = bind_context.result_columns();
```
Comment (leiysky) [inline query/src/sql/planner/binder/scalar.rs:169]: Maybe use `BindContext::all_column_bindings()`?

File: query/src/sql/planner/binder/scalar.rs [hunk 2]
```diff
@@ -160,6 +132,80 @@ impl ScalarBinder {
...
+        } else {
+            Scalar::AggregateFunction {
+                func_name: func_name.name.clone(),
+                distinct,
+                params: data_values,
+                args: scalar_exprs,
+                data_type: agg_func_ref.return_type()?,
+                nullable: false,
+            }
+        };
```
Comment (leiysky) [inline query/src/sql/planner/binder/scalar.rs:206]:
```suggestion
        let agg_scalar = Scalar::AggregateFunction {
            func_name: func_name.name.clone(),
            distinct,
            params: data_values,
            args: if optimize_remove_count_args(func_name.name.as_str(), distinct, args) { vec![] } else { scalar_exprs },
            data_type: agg_func_ref.return_type()?,
            nullable: false,
        };
```

Comment (xudong963) [reply query/src/sql/planner/binder/scalar.rs:206]: Nice, It is surprisingly possible to put `if else` here 😲

File: query/src/sql/planner/binder/scalar.rs [hunk 3]
```diff
@@ -160,6 +132,68 @@ impl ScalarBinder {
...
+        }
+
+        let scalar_binder = ScalarBinder::new();
+        let mut scalar_exprs = Vec::with_capacity(args.len());
+        for arg in args.iter() {
+            scalar_exprs.push(
+                scalar_binder
+                    .bind_expr(arg, bind_context)?
+                    .as_any()
+                    .downcast_ref::<Scalar>()
```
Comment (BohuTANG) [inline query/src/sql/planner/binder/scalar.rs:163]: We can use `enum-dispatch` instead of `downcast_ref`, see https://github.com/datafuselabs/databend/pull/5077
That's ok to do it in another PR :D

Comment (xudong963) [reply query/src/sql/planner/binder/scalar.rs:163]: I had tried in my local, but a blocking point is encountered due to https://github.com/datafuselabs/databend/blob/218821b4120e511ab9afacf25d0601a8ef90b340/query/src/sql/planner/binder/project.rs#L108

**Comment by xudong963** [comment]
The [commit](https://github.com/datafuselabs/databend/pull/5115/commits/bb277628476c4f6fd85afed730e5b0b22ec8fdad) fixed a bug:
```sql
mysql> select * from t1;
+------+------+
| a    | b    |
+------+------+
|    1 |    2 |
|    2 |    3 |
|    3 |    4 |
+------+------+
3 rows in set (0.05 sec)
Read 3 rows, 24 B in 0.021 sec., 140.25 rows/sec., 1.12 KB/sec.

mysql> SELECT sum(a) FROM t1 group by a;
ERROR 1105 (HY000): Code: 1028, displayText = sum expect to have single arguments, but got 2.
mysql>
```
Now it normally works!

---

Btw, the following is todos for aggregator in new planner:

- [x] count(*), count(col), count(1)
- select sum(a) from t;
- select count(a) from t;
- select count() from t group by a;
- select count() from t;
- select count(1) from t;
- select count(1) from t group by a;
- [x] There is a bug when the table contains multiple columns
- [ ] support having
- [ ] support scalar expression, such as a + 1, b + 1 in group by
- [ ] https://dev.mysql.com/doc/refman/8.0/en/group-by-handling.html,  the aggregation function supports the inclusion of columns that are not in the group by: low pri

I will open new tickets for the rest todos ~

**Comment by leiysky** [comment]
What about moving the aggregation related code to a new file `aggregate.rs`?

**Comment by xudong963** [comment]
> What about moving the aggregation related code to a new file `aggregate.rs`?

done

File: query/src/sql/planner/binder/scalar.rs (hunk 3 above)
Comment (sundy-li) [reply query/src/sql/planner/binder/scalar.rs:167]: Yes, I tries but failed.

**Review by sundy-li — APPROVED** [review]
